### PR TITLE
mnist tutorial l2 typo

### PR DESCRIPTION
**Pull request by endremborza**
Hello!

ran into this in a tutorial and it was bugging me, and since github now makes it impressively easy to submit such a fix, I did.

I tried it during the tutorial and it actually improved accuracy a tiny bit :)

enjoyed the tutorial and jax+haiku though, thanks for sharing.

### CODE REVIEW

**Comment by endremborza** [comment]
@googlebot I signed it!

**Comment by ajbrock** [comment]
Hi Endre,

This is not a typo: the implementation of L2 regularization is correct, and there should not be a sqrt anywhere. The quantity which is being regularized is not actually the Euclidean norm (which would indeed have the sqrt) but the squared Euclidean norm.

An easy way to prove this to oneself is to recognize that the gradient of L2 parameter regularization is simply the value of the parameter itself. This is why, in frameworks like PyTorch where L2 reg is implemented as weight decay inside the optimizer, it takes the form:

`param = param - (grad + lambda * param)`

where `lambda` is the weight decay strength. Thus, the gradient of the L2 loss should be the parameter term. Consider this snippet to demonstrate:

```
import jax
import jax.numpy as jnp
def l2_nosqrt(x):
  return 0.5 * jnp.sum(x**2)
def l2(x):
  return jnp.sum(x**2)**0.5
l2_nosqrt_grad = jax.grad(l2_nosqrt)
l2_grad = jax.grad(l2)
key = jax.random.PRNGKey(0)
shape = (8, 8)
x = jax.random.normal(key, shape)
print(f'L2 with sqrt provides correct grads: {jnp.allclose(l2_grad(x), x)}')
print(f'L2 without sqrt provides correct grads: {jnp.allclose(l2_nosqrt_grad(x), x)}')
```
which yields:
```
L2 with sqrt provides correct grads: False
L2 without sqrt provides correct grads: True
```

Another way you can build intuition here is to recognize that if you apply the square root to the summed terms, the strength of regularization will vary with the size of x (e.g. the dimensionality of all the tensors being summed), because the square root is outside the enclosing sum. While you certainly can do this if you want (other techniques like global gradient norm clipping also have their effective strength vary with the dimensionality of the parameters), it would differ from the standard implementation.

This can also be seen historically if you look at [Lasagne's implementation](https://github.com/Lasagne/Lasagne/blob/master/lasagne/regularization.py#L70-L82) of L2 loss, or at [Tensorflow's Implementation](https://www.tensorflow.org/api_docs/python/tf/nn/l2_loss).

**Comment by endremborza** [comment]
Thank you for taking the time to explain so quickly. I looked into it a little further, it's a bit frustrating not being able to tell why this works in some cases, but I think I see the evolution of the convention now and it is at least certainly simpler.
